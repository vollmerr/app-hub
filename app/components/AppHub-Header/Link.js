import React from 'react';
import styled from 'styled-components';
import { CommandButton, ElementType } from 'office-ui-fabric-react/lib/Button';
import { Link as RouterLink } from 'react-router-dom';

import theme from 'utils/theme';
import PropTypes from 'prop-types';
// import HeaderItem from 'components/Header/HeaderItem';
import Item from './Item';

const LinkStyled = styled(RouterLink) `
  text-decoration: none;
`;

const A = LinkStyled.withComponent('a');

const ButtonStyled = styled(CommandButton) `
  color: ${theme.themePrimary};
  display: block;
  vertical-align: middle;
  border-width: 0;
  margin: 0;
  padding: 0 15px;
  min-width: ${theme.hub.headerHeight};
  height: ${theme.hub.headerHeight};
  box-sizing: border-box;
  text-align: center;
  line-height: normal;
  overflow: visible;
  transition: background-color .467s cubic-bezier(.1,.9,.2,1) 34ms;
  pointer-events: all;
  font-size: ${(props) => props.size || '28px'};
`;

class Link extends React.PureComponent {
  render() {
    const { panel, title, iconProps, onClick, to, href, text, children } = this.props;

    const isLink = onClick || to || href;

    const Button = () => (
      <ButtonStyled
        onClick={onClick ? () => onClick(panel) : null}
        iconProps={iconProps}
        title={title}
        ariaLabel={title}
      >
        {text}
        {children}
      </ButtonStyled>
    );

    // handle internal routing
    if (to) {
      return (
        <Item isLink={isLink}>
          <LinkStyled to={to} >
            <Button />
          </LinkStyled>
        </Item>
      );
    } else if (href) {
      return (
        <Item isLink={isLink}>
          <A href={href} >
            <Button />
          </A>
        </Item>
      );
    }

    // no links
    return (
      <Item isLink={isLink}>
        <Button />
      </Item>
    );
  }
}

Link.propTypes = {
  panel: PropTypes.string,
  title: PropTypes.string,
  iconProps: PropTypes.object,
  onClick: PropTypes.func,
  text: PropTypes.string,
  to: PropTypes.string,
  href: PropTypes.string,
  children: PropTypes.node,
};

export default Link;
