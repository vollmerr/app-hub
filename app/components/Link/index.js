/**
*
* Link
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import { Link as RouterLink } from 'react-router-dom';
import styled from 'styled-components';

export const StyledLink = styled(RouterLink) `
  text-decoration: none;
`;

export const A = styled.a`
  text-decoration: none;
`;

class Link extends React.PureComponent {
  render() {
    const { to, href, children, ...props } = this.props;

    if (to) {
      return (
        <StyledLink to={to} {...props}>
          {children}
        </StyledLink>
      );
    }

    return (
      <A href={href} target={'_blank'} rel={'noopener noreferrer'} {...props}>
        {children}
      </A>
    );
  }
}

Link.propTypes = {
  to: PropTypes.string,
  href: PropTypes.string,
  children: PropTypes.node,
};

export default Link;
