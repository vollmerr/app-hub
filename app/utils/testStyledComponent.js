import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

function testStyledComponent(Component, extend = null) {
  describe(`<${Component.displayName} />`, () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<Component />);
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it(`should render an <${Component.target}> tag`, () => {
      expect(wrapper.type()).toEqual(Component.target);
    });

    it('should have a className attribute', () => {
      expect(wrapper.prop('className')).toBeDefined();
    });

    it('should adopt a valid attribute', () => {
      const id = 'test';
      wrapper.setProps({ id });
      expect(wrapper.prop('id')).toEqual(id);
    });

    if (extend) {
      it(`should extend ${extend.name}`, () => {
        expect(wrapper.find(extend).exists()).toEqual(true);
      });
    } else {
      it('should not adopt an invalid attribute', () => {
        const attribute = 'test';
        wrapper.setProps({ attribute });
        expect(wrapper.prop('attribute')).toBeUndefined();
      });
    }
  });
}

export default testStyledComponent;

