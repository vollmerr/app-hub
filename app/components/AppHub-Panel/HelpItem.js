import React from 'react';
import styled from 'styled-components';

import Link from 'components/Link';
import theme from 'utils/theme';

export const Item = styled.li`
  list-style-type: none;
  margin: 0 -20px;

  &:hover {
    background: ${theme.neutralLighter};
  }
`;

export const StlyedLink = styled(Link) `
  color: ${theme.white};
  display: block;
  padding: 15px;

  &:hover,
  &:active,
  &:focus {
    color: ${theme.neutralDark};
  }
`;

class HelpItem extends React.PureComponent {
  render() {
    return (
      <Item>
        <StlyedLink {...this.props} />
      </Item>
    );
  }
}

export default HelpItem;
