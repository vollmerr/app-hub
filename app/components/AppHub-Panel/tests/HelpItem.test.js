import React from 'react';
import { shallow } from 'enzyme';

import testStyledComponent from 'utils/testStyledComponent';

import Link from 'components/Link';
import HelpItem, { StlyedLink, Item } from '../HelpItem';

testStyledComponent(StlyedLink, Link);
testStyledComponent(Item);


describe('<HelpItem />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<HelpItem />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should contain a link', () => {
    expect(wrapper.find(StlyedLink).length).toEqual(1);
  });

  it('should pass props to the link', () => {
    const to = '/test-route';
    wrapper.setProps({ to });
    expect(wrapper.find(StlyedLink).prop('to')).toEqual(to);
  });
});
