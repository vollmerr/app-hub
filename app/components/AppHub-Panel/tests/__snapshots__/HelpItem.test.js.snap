// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HelpItem /> should render correctly 1`] = `
<HelpItem__Item>
  <HelpItem__StlyedLink />
</HelpItem__Item>
`;

exports[`<HelpItem__Item /> should render correctly 1`] = `
.c0 {
  list-style-type: none;
  margin: 0 -20px;
}

.c0:hover {
  background: #f4f4f4;
}

<li
  className="c0"
/>
`;

exports[`<HelpItem__StlyedLink /> should render correctly 1`] = `
.c0 {
  color: #ffffff;
  display: block;
  padding: 15px;
}

.c0:hover,
.c0:active,
.c0:focus {
  color: #212121;
}

<Link
  className="c0"
/>
`;
